cmd/remote_val_harness/internal/test_harness.go

```diff
@@ -0,0 +1,391 @@
+		th.logger.Info("Testing vote type", "type", voteType)
+		hash := tmhash.Sum([]byte("hash"))
+		vote := &types.Vote{
+			Type:   voteType,
+			Height: 12345,
+			Round:  23456,
+			BlockID: types.BlockID{
+				Hash: hash,
+				PartsHeader: types.PartSetHeader{
+					Hash:  hash,
+					Total: 1000000,
+				},
+			},
+			ValidatorIndex:   0,
+			ValidatorAddress: tmhash.SumTruncated([]byte("addr")),
+			Timestamp:        time.Now(),
+		}
+		voteBytes := vote.SignBytes(th.chainID)
+		// sign the vote
+		if err := th.sc.SignVote(th.chainID, vote); err != nil {
+			th.logger.Error("FAILED: Signing of vote", "err", err)
+			return newTestHarnessError(ErrTestSignVoteFailed, err, fmt.Sprintf("voteType=%d", voteType))
+		}
+		th.logger.Debug("Signed vote", "vote", vote)
+		// validate the contents of the vote
+		if err := vote.ValidateBasic(); err != nil {
+			th.logger.Error("FAILED: Signed vote is invalid", "err", err)
+			return newTestHarnessError(ErrTestSignVoteFailed, err, fmt.Sprintf("voteType=%d", voteType))
+		}
+		// now validate the signature on the proposal
+		if th.sc.GetPubKey().VerifyBytes(voteBytes, vote.Signature) {
+			th.logger.Info("Successfully validated vote signature", "type", voteType)
+		} else {
+			th.logger.Error("FAILED: Vote signature validation failed", "type", voteType)
+			return newTestHarnessError(ErrTestSignVoteFailed, nil, "signature validation failed")
+		}
+	}
+	return nil
+}
+
+// Shutdown will kill the test harness and attempt to close all open sockets
+// gracefully. If the supplied error is nil, it is assumed that the exit code
+// should be 0. If err is not nil, it will exit with an exit code related to the
+// error.
+func (th *TestHarness) Shutdown(err error) {
+	var exitCode int
+
+	if err == nil {
+		exitCode = NoError
+	} else if therr, ok := err.(*TestHarnessError); ok {
+		exitCode = therr.Code
+	} else {
+		exitCode = ErrOther
+	}
+	th.exitCode = exitCode
+
+	// in case sc.Stop() takes too long
+	if th.exitWhenComplete {
+		go func() {
+			time.Sleep(time.Duration(5) * time.Second)
+			th.logger.Error("Forcibly exiting program after timeout")
+			os.Exit(exitCode)
+		}()
+	}
+
+	if th.sc.IsRunning() {
+		if err := th.sc.Stop(); err != nil {
+			th.logger.Error("Failed to cleanly stop listener: %s", err.Error())
+		}
+	}
+
+	if th.exitWhenComplete {
+		os.Exit(exitCode)
+	}
+}
+
+// newTestHarnessSocketVal creates our client instance which we will use for
+// testing.
+func newTestHarnessSocketVal(logger log.Logger, cfg TestHarnessConfig) (*privval.SocketVal, error) {
+	proto, addr := cmn.ProtocolAndAddress(cfg.BindAddr)
+	if proto == "unix" {
+		// make sure the socket doesn't exist - if so, try to delete it
+		if cmn.FileExists(addr) {
+			if err := os.Remove(addr); err != nil {
+				logger.Error("Failed to remove existing Unix domain socket", "addr", addr)
+				return nil, err
+			}
+		}
+	}
+	ln, err := net.Listen(proto, addr)
+	logger.Info("Listening at", "proto", proto, "addr", addr)
```

Comment:
Maybe move this line underneath the error check below?

Reply:
Yes, good catch 😊 👍